File: vendor_app/lib/models/order.dart
enum OrderStatus {
  pending,
  accepted,
  preparing,
  ready,
  outForDelivery,
  delivered,
  completed,
  cancelled,
}

extension OrderStatusX on OrderStatus {
  String get label {
    switch (this) {
      case OrderStatus.pending:
        return 'Pending';
      case OrderStatus.accepted:
        return 'Accepted';
      case OrderStatus.preparing:
        return 'Preparing';
      case OrderStatus.ready:
        return 'Ready';
      case OrderStatus.outForDelivery:
        return 'Out for Delivery';
      case OrderStatus.delivered:
        return 'Delivered';
      case OrderStatus.completed:
        return 'Completed';
      case OrderStatus.cancelled:
        return 'Cancelled';
    }
  }

  String get key {
    switch (this) {
      case OrderStatus.pending:
        return 'pending';
      case OrderStatus.accepted:
        return 'accepted';
      case OrderStatus.preparing:
        return 'preparing';
      case OrderStatus.ready:
        return 'ready';
      case OrderStatus.outForDelivery:
        return 'out_for_delivery';
      case OrderStatus.delivered:
        return 'delivered';
      case OrderStatus.completed:
        return 'completed';
      case OrderStatus.cancelled:
        return 'cancelled';
    }
  }

  /// The next logical status a vendor can move an order to. Pickup orders
  /// skip the delivery-tracking flow entirely (ready → collected =
  /// completed), matching the backend's guarded transitions.
  OrderStatus? nextFor(FulfillmentType fulfillmentType) {
    switch (this) {
      case OrderStatus.pending:
        return OrderStatus.accepted;
      case OrderStatus.accepted:
        return OrderStatus.preparing;
      case OrderStatus.preparing:
        return OrderStatus.ready;
      case OrderStatus.ready:
        return fulfillmentType == FulfillmentType.pickup
            ? OrderStatus.completed
            : OrderStatus.outForDelivery;
      case OrderStatus.outForDelivery:
        return OrderStatus.delivered;
      case OrderStatus.delivered:
        return OrderStatus.completed;
      case OrderStatus.completed:
      case OrderStatus.cancelled:
        return null;
    }
  }
}

enum FulfillmentType { delivery, pickup, scheduled }

class OrderLineItem {
  final String foodName;
  final int quantity;
  final double price;
  final List<String> addons;
  final String specialInstructions;

  OrderLineItem({
    required this.foodName,
    required this.quantity,
    required this.price,
    this.addons = const [],
    this.specialInstructions = '',
  });

  double get subtotal => price * quantity;
}

class VendorOrder {
  final String id;
  final String customerName;
  final String customerMobile;
  final String deliveryAddress;
  final double? customerLat;
  final double? customerLng;
  final List<OrderLineItem> items;
  OrderStatus status;
  final FulfillmentType fulfillmentType;
  final DateTime placedAt;
  final DateTime? scheduledFor;
  final double deliveryFee;
  final String paymentMethod;
  final String notes; // order-level note, e.g. "Leave at the gate, thanks!"
  final String orderNumber;

  VendorOrder({
    required this.id,
    required this.customerName,
    required this.customerMobile,
    required this.deliveryAddress,
    this.customerLat,
    this.customerLng,
    required this.items,
    this.status = OrderStatus.pending,
    this.fulfillmentType = FulfillmentType.delivery,
    required this.placedAt,
    this.scheduledFor,
    this.deliveryFee = 0,
    this.paymentMethod = 'Cash on Delivery',
    this.notes = '',
    this.orderNumber = '',
  });

  double get itemsTotal => items.fold(0, (sum, i) => sum + i.subtotal);
  double get grandTotal => itemsTotal + deliveryFee;

  static OrderStatus _statusFromKey(String? key) =>
      OrderStatus.values.firstWhere(
        (s) => s.key == key,
        orElse: () => OrderStatus.pending,
      );

  static FulfillmentType _fulfillmentFromKey(String? key) {
    switch (key) {
      case 'pickup':
        return FulfillmentType.pickup;
      case 'scheduled_delivery':
        return FulfillmentType.scheduled;
      default:
        return FulfillmentType.delivery;
    }
  }

  /// Maps the backend's OrderOut (vendor inbox shape, snake_case).
  factory VendorOrder.fromApi(Map<String, dynamic> json) {
    DateTime? parseDate(String? raw) =>
        raw == null ? null : DateTime.tryParse(raw)?.toLocal();
    return VendorOrder(
      id: json['id'] as String,
      orderNumber: json['order_number'] as String? ?? '',
      customerName: json['customer_name'] as String? ?? 'Customer',
      customerMobile: json['customer_mobile'] as String? ?? '',
      deliveryAddress: json['delivery_address'] as String? ?? '',
      customerLat: (json['delivery_latitude'] as num?)?.toDouble(),
      customerLng: (json['delivery_longitude'] as num?)?.toDouble(),
      items: [
        for (final item in (json['items'] as List?) ?? [])
          OrderLineItem(
            foodName: (item as Map<String, dynamic>)['item_name'] as String? ?? '',
            quantity: item['quantity'] as int? ?? 1,
            price: (item['unit_price'] as num?)?.toDouble() ?? 0,
            addons: [
              for (final choices
                  in ((item['selected_options'] as Map?)?.cast<String, dynamic>() ?? const {})
                      .values)
                ...((choices as List?)?.cast<String>() ?? const []),
            ],
            specialInstructions:
                item['special_instructions'] as String? ?? '',
          ),
      ],
      status: _statusFromKey(json['status'] as String?),
      fulfillmentType: _fulfillmentFromKey(json['delivery_method'] as String?),
      placedAt: parseDate(json['created_at'] as String?) ?? DateTime.now(),
      scheduledFor: parseDate(json['scheduled_for'] as String?),
      deliveryFee: (json['delivery_fee'] as num?)?.toDouble() ?? 0,
      paymentMethod: paymentLabel(json['payment_method'] as String?),
      notes: json['special_instructions'] as String? ?? '',
    );
  }

  /// Human label for every PaymentMethod the backend can send.
  static String paymentLabel(String? key) {
    switch (key) {
      case 'cash_on_pickup':
        return 'Cash on Pickup';
      case 'gcash':
        return 'GCash';
      case 'maya':
        return 'Maya';
      case 'credit_card':
        return 'Credit/Debit Card';
      case 'bank_transfer':
        return 'Bank Transfer';
      default:
        return 'Cash on Delivery';
    }
  }
}

File: backend/app/schemas/order.py
import uuid
from datetime import datetime

from pydantic import BaseModel, model_validator

from app.models.enums import DeliveryMethod, OrderStatus, PaymentMethod


class CartItemIn(BaseModel):
    food_item_id: uuid.UUID
    quantity: int = 1
    selected_options: dict[str, list[str]] = {}  # {"Extras": ["Add Cheese"]}
    special_instructions: str | None = None


class CheckoutRequest(BaseModel):
    vendor_id: uuid.UUID
    items: list[CartItemIn]
    delivery_method: DeliveryMethod
    payment_method: PaymentMethod
    address_id: uuid.UUID | None = None  # required if delivery_method != pickup
    scheduled_for: datetime | None = None  # required if delivery_method == scheduled_delivery
    special_instructions: str | None = None
    promo_code: str | None = None

    @model_validator(mode="after")
    def validate_method_requirements(self):
        if self.delivery_method != DeliveryMethod.pickup and not self.address_id:
            raise ValueError("address_id is required for delivery orders")
        if self.delivery_method == DeliveryMethod.scheduled_delivery and not self.scheduled_for:
            raise ValueError("scheduled_for is required for scheduled deliveries")
        return self


class PromoValidateRequest(BaseModel):
    vendor_id: uuid.UUID
    code: str
    subtotal: float


class PromoValidateOut(BaseModel):
    valid: bool
    title: str | None = None
    discount: float = 0
    message: str | None = None


class OrderItemOut(BaseModel):
    id: uuid.UUID
    item_name: str
    unit_price: float
    quantity: int
    selected_options: dict
    line_total: float
    special_instructions: str | None

    class Config:
        from_attributes = True


class OrderOut(BaseModel):
    id: uuid.UUID
    order_number: str
    vendor_id: uuid.UUID
    status: OrderStatus
    delivery_method: DeliveryMethod
    payment_method: PaymentMethod
    subtotal: float
    delivery_fee: float
    discount: float = 0
    total: float
    scheduled_for: datetime | None
    created_at: datetime
    items: list[OrderItemOut]
    # Populated for the vendor inbox (customer-facing endpoints omit them):
    customer_name: str | None = None
    customer_mobile: str | None = None
    delivery_address: str | None = None
    delivery_latitude: float | None = None
    delivery_longitude: float | None = None
    special_instructions: str | None = None
    cancellation_reason: str | None = None

    class Config:
        from_attributes = True


class UpdateOrderStatus(BaseModel):
    status: OrderStatus
    note: str | None = None
